File: src/game/PlayerManager.js
/**
 * PlayerManager handles player setup and validation
 * Enforces rules for player count (2-8), unique names, and game start validation
 */
class PlayerManager {
  constructor() {
    this.players = [];
    this.MAX_PLAYERS = 8;
    this.MIN_PLAYERS = 2;
  }

  /**
   * Add a player with validation
   * @param {string} name - Player name
   * @returns {object} Result object with success flag and message
   * @throws {Error} If validation fails
   */
  addPlayer(name) {
    // Trim and validate name
    const trimmedName = name.trim();
    
    if (!trimmedName) {
      return {
        success: false,
        message: 'Player name cannot be empty'
      };
    }

    // Check for duplicate names (case-insensitive)
    const nameLowerCase = trimmedName.toLowerCase();
    const isDuplicate = this.players.some(
      player => player.name.toLowerCase() === nameLowerCase
    );
    
    if (isDuplicate) {
      return {
        success: false,
        message: `Player "${trimmedName}" already exists`
      };
    }

    // Check player count limit
    if (this.players.length >= this.MAX_PLAYERS) {
      return {
        success: false,
        message: `Maximum of ${this.MAX_PLAYERS} players allowed`
      };
    }

    // Add the player
    const newPlayer = {
      id: this.generatePlayerId(),
      name: trimmedName,
      addedAt: new Date()
    };
    
    this.players.push(newPlayer);
    
    return {
      success: true,
      message: `Player "${trimmedName}" added successfully`,
      player: newPlayer
    };
  }

  /**
   * Remove a player from the setup
   * @param {string|number} playerIdOrName - Player ID or name
   * @returns {object} Result object with success flag and message
   */
  removePlayer(playerIdOrName) {
    const initialLength = this.players.length;
    
    this.players = this.players.filter(player => {
      // Match by ID (number) or name (string)
      if (typeof playerIdOrName === 'number') {
        return player.id !== playerIdOrName;
      }
      return player.name.toLowerCase() !== playerIdOrName.toLowerCase();
    });

    if (this.players.length < initialLength) {
      return {
        success: true,
        message: 'Player removed successfully'
      };
    }
    
    return {
      success: false,
      message: 'Player not found'
    };
  }

  /**
   * Get all current players
   * @returns {array} Array of player objects
   */
  getPlayers() {
    return [...this.players];
  }

  /**
   * Get player count
   * @returns {number} Current number of players
   */
  getPlayerCount() {
    return this.players.length;
  }

  /**
   * Check if game can start
   * @returns {object} Validation result with canStart flag and message
   */
  canStartGame() {
    if (this.players.length < this.MIN_PLAYERS) {
      return {
        canStart: false,
        message: `Minimum of ${this.MIN_PLAYERS} players required to start game`,
        playerCount: this.players.length,
        minRequired: this.MIN_PLAYERS
      };
    }

    if (this.players.length > this.MAX_PLAYERS) {
      return {
        canStart: false,
        message: `Too many players. Maximum is ${this.MAX_PLAYERS}`,
        playerCount: this.players.length,
        maxAllowed: this.MAX_PLAYERS
      };
    }

    return {
      canStart: true,
      message: 'Game can start',
      playerCount: this.players.length
    };
  }

  /**
   * Reset the player setup
   */
  reset() {
    this.players = [];
  }

  /**
   * Generate a unique player ID
   * @private
   * @returns {number} Unique player ID
   */
  generatePlayerId() {
    if (this.players.length === 0) {
      return 1;
    }
    return Math.max(...this.players.map(p => p.id)) + 1;
  }

  /**
   * Check if a name is available
   * @param {string} name - Player name to check
   * @returns {boolean} True if name is available
   */
  isNameAvailable(name) {
    const nameLowerCase = name.trim().toLowerCase();
    return !this.players.some(
      player => player.name.toLowerCase() === nameLowerCase
    );
  }
}

module.exports = PlayerManager;

File: lib/managers/player_manager.dart
import 'package:uuid/uuid.dart';
import '../models/player.dart';

/// Exception thrown when player operations fail
class PlayerException implements Exception {
  final String message;
  PlayerException(this.message);

  @override
  String toString() => message;
}

/// Manages player operations during game setup
class PlayerManager {
  static const int minPlayers = 2;
  static const int maxPlayers = 8;
  static const _uuid = Uuid();

  final List<Player> _players = [];

  /// Returns the current list of players
  List<Player> get players => List.unmodifiable(_players);

  /// Returns the current number of players
  int get playerCount => _players.length;

  /// Checks if a player name already exists (case-insensitive)
  bool _playerNameExists(String name) {
    final normalizedName = name.trim().toLowerCase();
    return _players.any((p) => p.name.toLowerCase() == normalizedName);
  }

  /// Adds a new player to the game
  ///
  /// Throws [PlayerException] if:
  /// - The name is invalid
  /// - A player with that name already exists
  /// - The maximum number of players is reached
  Player addPlayer(String name) {
    // Validate name format
    final validationError = Player.validateName(name);
    if (validationError != null) {
      throw PlayerException(validationError);
    }

    final trimmedName = name.trim();

    // Check for duplicate names (case-insensitive)
    if (_playerNameExists(trimmedName)) {
      throw PlayerException(
          'A player with the name "$trimmedName" already exists');
    }

    // Check player limit
    if (_players.length >= maxPlayers) {
      throw PlayerException(
          'Cannot add more than $maxPlayers players');
    }

    final newPlayer = Player(
      id: _uuid.v4(),
      name: trimmedName,
    );

    _players.add(newPlayer);
    return newPlayer;
  }

  /// Removes a player by their ID
  ///
  /// Throws [PlayerException] if the player is not found
  void removePlayer(String playerId) {
    final index = _players.indexWhere((p) => p.id == playerId);
    if (index == -1) {
      throw PlayerException('Player with ID "$playerId" not found');
    }
    _players.removeAt(index);
  }

  /// Removes a player by their name (case-insensitive)
  ///
  /// Throws [PlayerException] if the player is not found
  void removePlayerByName(String name) {
    final normalizedName = name.trim().toLowerCase();
    final index = _players.indexWhere(
        (p) => p.name.toLowerCase() == normalizedName);
    if (index == -1) {
      throw PlayerException('Player with name "$name" not found');
    }
    _players.removeAt(index);
  }

  /// Gets a player by their ID
  Player? getPlayer(String playerId) {
    try {
      return _players.firstWhere((p) => p.id == playerId);
    } catch (e) {
      return null;
    }
  }

  /// Gets a player by their name (case-insensitive)
  Player? getPlayerByName(String name) {
    final normalizedName = name.trim().toLowerCase();
    try {
      return _players.firstWhere(
          (p) => p.name.toLowerCase() == normalizedName);
    } catch (e) {
      return null;
    }
  }

  /// Checks if the game can start
  ///
  /// Returns true if there are at least [minPlayers] players
  bool canStartGame() => _players.length >= minPlayers;

  /// Validates that the game can start
  ///
  /// Throws [PlayerException] if there are fewer than [minPlayers] players
  void validateGameStart() {
    if (!canStartGame()) {
      throw PlayerException(
          'At least $minPlayers players are required to start the game');
    }
  }

  /// Clears all players (useful for resetting or testing)
  void clear() {
    _players.clear();
  }
}
